File: 49-group-anagrams/group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        hashMap = defaultdict(list) # Mapping character count to list of anagrams

        for string in strs:
            count = [0] * 26 # a . . . z

            for char in string:
                count[ord(char) - ord("a")] += 1

            hashMap[tuple(count)].append(string)

        return list(hashMap.values())

File: 49-group-anagrams/group-anagrams.cs
public class Solution {
    public IList<IList<string>> GroupAnagrams(string[] strs) {
        var hashMap = new Dictionary<string, List<string>>();
        
        foreach (var str in strs) {
            // Count characters
            int[] count = new int[26];
            foreach (char c in str) {
                count[c - 'a']++;
            }

            // Build a key string from count array
            string key = string.Join("#", count);

            // Add to dictionary
            if (!hashMap.ContainsKey(key)) {
                hashMap[key] = new List<string>();
            }
            hashMap[key].Add(str);
        }

        // Convert dictionary values to result list
        var result = new List<IList<string>>();
        foreach (var group in hashMap.Values) {
            result.Add(group);
        }
        
        return result;
    }
}